File: Algorithms/wildcardMatching.py
class Solution:
    # @param {string} s
    # @param {string} p
    # @return {boolean}
    def isMatch(self, s, p):
        star, i = False, 0
        while i < len(s):
            if i >= len(p):
                if star:
                    s, i = s[1:], -1
                else: 
                    return False
            elif p[i] == '?' or s[i] == p[i]:
                pass
            elif p[i] == '*':
                star = True
                p, s, i = p[i+1:], s[i:], -1
                while p and p[0] == '*':
                    p = p[1:]
                if not p:
                    return True
            elif s[i] != p[i]:
                if star:
                    s, i = s[1:], -1
                else: 
                    return False
            i += 1
        while i < len(p) and p[i] == '*':
            i += 1
        return i == len(p)

File: Algorithms/wildcardMatching.java
public class Solution {
    public boolean isMatch(String s, String p) {
        int m = s.length();
        int mm = 0;
        int n = p.length();
        int nn = 0;
        boolean star = false;
        int i;
        for (i=0; i+mm < m; i++) {
            if (i+nn >= n) {
                if (star) {
                    mm ++;
                    i = -1;
                    continue;
                } else {
                    return false;
                }
            }
            switch(p.charAt(i+nn)) {
                case '?':
                    break;
                case '*':
                    star = true;
                    nn += i+1;
                    mm += i;
                    i = 0;
                    while (i+nn < n && p.charAt(i+nn) == '*') {
                        i ++;
                    }
                    if (i+nn == n) 
                        return true;
                    i = -1;
                    break;
                default:
                    if (s.charAt(i+mm) != p.charAt(i+nn)) {
                        if (star) {
                            mm ++;
                            i = -1;
                        } else {
                            return false;
                        }
                    }
                    break;
            }
        }
        while (i+nn < n && p.charAt(i+nn) == '*') {
            i ++;
        }
        return i+nn == n;
    }
}
